src/resources/espnow.cc

```diff
@@ -0,0 +1,277 @@
+};
+
+const int kInvalidESPNow = -1;
+
+// Only allow one instance of provisioning running.
+static  ResourcePool<int, kInvalidESPNow> espnow_pool(
+  0
+);
+
+static SemaphoreHandle_t tx_sem;
+static esp_now_send_status_t tx_status;
+static QueueHandle_t rx_queue;
+static SemaphoreHandle_t rx_datagrams_mutex;
+static struct DataGram* rx_datagrams;
+
+class ESPNowResourceGroup : public ResourceGroup {
+ public:
+  TAG(ESPNowResourceGroup);
+
+  ESPNowResourceGroup(Process* process);
+  ~ESPNowResourceGroup();
+};
+
+ESPNowResourceGroup::ESPNowResourceGroup(Process* process)
+  : ResourceGroup(process) {
+  tx_sem = xSemaphoreCreateCounting(1, 0);
+  ASSERT(tx_sem);
+
+  rx_queue = xQueueCreate(ESPNOW_RX_DATAGRAM_NUM, sizeof(void *));
+  ASSERT(rx_queue);
+
+  rx_datagrams_mutex = xSemaphoreCreateMutex();
+  ASSERT(rx_datagrams_mutex);
+
+  rx_datagrams = unvoid_cast<struct DataGram*>(malloc(sizeof(struct DataGram) * ESPNOW_RX_DATAGRAM_NUM));
+  ASSERT(rx_datagrams);
+}
+
+ESPNowResourceGroup::~ESPNowResourceGroup() {
+  vSemaphoreDelete(tx_sem);
+  tx_sem = NULL;
+
+  vQueueDelete(rx_queue);
+  rx_queue = NULL;
+
+  vSemaphoreDelete(rx_datagrams_mutex);
+  rx_datagrams_mutex = NULL;
+
+  free(rx_datagrams);
+  rx_datagrams = NULL;
+}
+
+MODULE_IMPLEMENTATION(espnow, MODULE_ESPNOW)
+
+static struct DataGram* alloc_datagram(void) {
+  struct DataGram *datagram = NULL;
+
+  xSemaphoreTake(rx_datagrams_mutex, portMAX_DELAY);
+  for (int i = 0; i < ESPNOW_RX_DATAGRAM_NUM; i++) {
+    if (!rx_datagrams[i].used) {
+      datagram = &rx_datagrams[i];
+      datagram->used = true;
+      break;
+    }
+  }
+  xSemaphoreGive(rx_datagrams_mutex);
+
+  return datagram;
+}
+
+static void free_datagram(struct DataGram* datagram) {
+  xSemaphoreTake(rx_datagrams_mutex, portMAX_DELAY);
+  datagram->used = false;
+  xSemaphoreGive(rx_datagrams_mutex);
+}
+
+static void espnow_send_cb(const uint8_t *mac_addr, esp_now_send_status_t status) {
+  tx_status = status;
+  portBASE_TYPE ret = xSemaphoreGive(tx_sem);
+  if (ret != pdTRUE) {
+    // ESP_LOGE("ESPNow", "Failed to give TX semaphore");
+    return ;
+  }
+}
+
+static void espnow_recv_cb(const uint8_t *mac_addr, const uint8_t *data, int data_len) {
+  if (data_len > ESPNOW_RX_DATAGRAM_LEN_MAX) {
+    ESP_LOGE("ESPNow", "Receive datagram length=%d is larger than max=%d", data_len, ESPNOW_RX_DATAGRAM_LEN_MAX);
+    return ;
+  }
+
+  struct DataGram* datagram = alloc_datagram();
+  if (!datagram) {
+    // ESP_LOGE("ESPNow", "Failed to malloc datagram");
+    return ;
+  }
+
+  datagram->len = data_len;
+  memcpy(datagram->mac, mac_addr, 6);
+  memcpy(datagram->buffer, data, data_len);
+
+  portBASE_TYPE ret = xQueueSend(rx_queue, &datagram, 0);
+  if (ret != pdTRUE) {
+    free_datagram(datagram);
+    // ESP_LOGE("ESPNow", "Failed to send datagram to rx_queue");
+  }
+}
+
+PRIMITIVE(init) {
+  ARGS(int, mode, Blob, pmk);
+
+  ByteArray* proxy = process->object_heap()->allocate_proxy();
```

Comment:
Add a check that the key is of the right size. We do that in Toit before, but it's still good practice to ensure your invariants in C++ again.

Reply:
Of source.